verification_service/main.py

```diff
@@ -0,0 +1,309 @@
+
+        for domain in verified_domains:
+            if re.search(domain, email):
+                return True
+    except (FileNotFoundError, yaml.YAMLError, KeyError) as e:
+        print(f"Error: {e}")
+
+    return False
+
+
+def check_user_verification(email, verified_users_file):
+    with open(verified_users_file, "r") as file:
+        verified_users = yaml.safe_load(file)["users"]
+
+    for verified_user in verified_users:
+        if verified_user == email:
+            return True
+    return False
+
+
+def get_user_email_address(event_type, event_data):
+    if event_type in project_events:
+        return event_data.get("owner_email")
+
+    elif event_type in user_events:
+        return event_data.get("email")
+
+    elif event_type in issue_events + issue_note_events:
+        user_attributes = event_data.get("user", {})
+        return user_attributes.get("email")
+
+    else:
+        logging.debug(
+            f"Unable to get user email address for this event type: {event_type}"
+        )
+        return None
+
+
+# VerificationEventProcessor class, which inherits from EventProcessor.
+# It is used to process events received from redis and push them back
+# into redis after processing, if the user or their email domain is
+# not verified.
+class VerificationEventProcessor(EventProcessor):
+    def __init__(
+        self,
+        prefix,
+        redis_conn=None,
+        verified_users_file=None,
+        verified_domains_file=None,
+        gitlab_url=None,
+        gitlab_access_token=None,
+    ):
+        super().__init__(prefix, event_types, redis_conn)
+        self.verified_users_file = verified_users_file
+        self.verified_domains_file = verified_domains_file
+        self.gitlab_url = gitlab_url
+        self.gitlab_access_token = gitlab_access_token
+
+    def process_event(self, queue_name, data):
+        event_type = queue_name.split("_", 1)[-1]
+
+        logging.debug(f"Processing event {event_type}")
+
+        processed_events_total.inc()
+
+        user_email_address = None
+        user_verified = False
+
+        # Determine how to get the user email address based on event type
+        if (
+            event_type
+            in project_events + user_events + issue_events + issue_note_events
+        ):
+            user_email_address = get_user_email_address(event_type, data)
+
+        # If the event is a group event, get the user email address by
+        # getting all members of the group, and then checking the
+        # verification status of the user with the highest access
+        # level in the group.
+        elif event_type in group_events:
+            logging.info(
+                f"{event_type} event type received, getting user email from GitLab API"
+            )
+            max_access_level = 0
+            user_id_with_max_access = None
+            group_id = data.get("group_id")
+
+            # Get all members of the group
+            response = requests.get(
```

Comment:
**nitpick (non-blocking):** we should probably use the gitlab client in this file as well, since we already use it in retrieval_service

Reply:
I will need to use the GitLab API to delete users when that service is written. This is because the GitLab client doesn't allow you to hard delete a user. So I should probably use the GitLab API in the retrieval service instead then.